stl/inc/ranges

```diff
@@ -486,6 +671,210 @@ namespace ranges {
+                // clang-format on
+                if constexpr (_Choice<_Rng>._Strategy == _St::_View) {
+                    return static_cast<_Rng&&>(_Range);
+                } else if constexpr (_Choice<_Rng>._Strategy == _St::_Ref) {
+                    return ref_view{static_cast<_Rng&&>(_Range)};
+                } else if constexpr (_Choice<_Rng>._Strategy == _St::_Subrange) {
+                    return subrange{static_cast<_Rng&&>(_Range)};
+                } else {
+                    static_assert(_Always_false<_Rng>, "Should be unreachable");
+                }
+            }
+        };
+
+        inline constexpr _All_fn all;
+
+        // ALIAS TEMPLATE views::all_t
+        template <viewable_range _Rng>
+        using all_t = decltype(all(_STD declval<_Rng>()));
+    } // namespace views
+
+    // CLASS TEMPLATE ranges::reverse_view
+    // clang-format off
+    template <view _Vw>
+        requires bidirectional_range<_Vw>
+    class reverse_view : public _Cached_position_t<!common_range<_Vw>, _Vw, reverse_view<_Vw>> {
+        // clang-format on
+    private:
+        /* [[no_unique_address]] */ _Vw _Base{};
+
+        template <class _Rng>
+        using _Rev_iter = reverse_iterator<iterator_t<_Rng>>;
+
+    public:
+        reverse_view() = default;
+        constexpr explicit reverse_view(_Vw _Base_) noexcept(is_nothrow_move_constructible_v<_Vw>) // strengthened
+            : _Base(_STD move(_Base_)) {}
+
+        _NODISCARD constexpr _Vw base() const& noexcept(
+            is_nothrow_copy_constructible_v<_Vw>) /* strengthened */ requires copy_constructible<_Vw> {
+            return _Base;
+        }
+        _NODISCARD constexpr _Vw base() && noexcept(is_nothrow_move_constructible_v<_Vw>) /* strengthened */ {
+            return _STD move(_Base);
+        }
+
+        _NODISCARD constexpr _Rev_iter<_Vw> begin() {
+            if constexpr (common_range<_Vw>) {
+                return _Rev_iter<_Vw>{_RANGES end(_Base)};
+            } else {
+                if (this->_Has_cache()) {
+                    return _Rev_iter<_Vw>{this->_Get_cache(_Base)};
+                }
+
+                iterator_t<_Vw> _First;
+                if constexpr (sized_range<_Vw>) {
+                    _First = _RANGES next(_RANGES begin(_Base), _RANGES distance(_Base));
+                } else {
+                    _First = _RANGES next(_RANGES begin(_Base), _RANGES end(_Base));
+                }
+                this->_Set_cache(_Base, _First);
+                return _Rev_iter<_Vw>{_STD move(_First)};
+            }
+        }
+
+        _NODISCARD constexpr auto begin() const noexcept(
+            noexcept(_Rev_iter<const _Vw>{_RANGES end(_Base)})) /* strengthened */ requires common_range<const _Vw> {
+            return _Rev_iter<const _Vw>{_RANGES end(_Base)};
+        }
+
+        _NODISCARD constexpr _Rev_iter<_Vw> end() noexcept(

```

Comment:
Why does this one not return `auto`?

Reply:
Here, for consistency with the working draft. In the working draft, the non-`const` overloads of `begin` and `end` name the return type because they can. The `const` overloads must return `auto` because `reverse_iterator<iterator_t<const _Vw>>` is ill-formed if `const _Vw` isn't a `range`.

Reply:
Consistency with the WP makes sense, thanks.